generic-contracts/src/main/java/com/scalar/dl/genericcontracts/table/v1_0_0/Scan.java

```diff
@@ -0,0 +1,310 @@
+package com.scalar.dl.genericcontracts.table.v1_0_0;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.JsonNodeType;
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.Multimaps;
+import com.scalar.dl.ledger.contract.JacksonBasedContract;
+import com.scalar.dl.ledger.database.AssetFilter;
+import com.scalar.dl.ledger.exception.ContractContextException;
+import com.scalar.dl.ledger.statemachine.Asset;
+import com.scalar.dl.ledger.statemachine.Ledger;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+public class Scan extends JacksonBasedContract {
+
+  @Nullable
+  @Override
+  public JsonNode invoke(
+      Ledger<JsonNode> ledger, JsonNode arguments, @Nullable JsonNode properties) {
+
+    // Get the table information
+    String tableName = arguments.get(Constants.QUERY_TABLE).asText();
+    String tableAssetId = getAssetIdForTable(tableName);
+    JsonNode table =
+        ledger
+            .get(tableAssetId)
+            .orElseThrow(() -> new ContractContextException(Constants.TABLE_NOT_EXIST))
+            .data();
+
+    // Scan records
+    ListMultimap<String, JsonNode> conditionsMap =
+        Multimaps.index(
+            arguments.get(Constants.QUERY_CONDITIONS),
+            condition -> condition.get(Constants.CONDITION_COLUMN).textValue());
+    if (isGetAvailable(table, conditionsMap)) {
+      return get(ledger, table, conditionsMap);
+    } else if (isIndexScanAvailable(table, conditionsMap)) {
+      return scan(ledger, table, conditionsMap);
+    } else {
+      throw new ContractContextException(Constants.INVALID_KEY_SPECIFICATION);
+    }
+  }
+
+  private boolean isGetAvailable(JsonNode table, ListMultimap<String, JsonNode> conditionsMap) {
+    String key = table.get(Constants.TABLE_KEY).textValue();
+    String keyType = table.get(Constants.TABLE_KEY_TYPE).textValue();
+    return conditionsMap.containsKey(key)
+        && conditionsMap.get(key).stream()
+            .anyMatch(condition -> isPrimaryKeyCondition(condition, keyType));
+  }
+
+  private boolean isIndexScanAvailable(
+      JsonNode table, ListMultimap<String, JsonNode> conditionsMap) {
+    for (JsonNode index : table.get(Constants.TABLE_INDEXES)) {
+      String indexKey = index.get(Constants.INDEX_KEY).asText();
+      String indexKeyType = index.get(Constants.INDEX_KEY_TYPE).asText();
+      if (conditionsMap.containsKey(indexKey)
+          && conditionsMap.get(indexKey).stream()
+              .anyMatch(condition -> isIndexKeyCondition(condition, indexKeyType))) {
+        return true;
+      }
+    }
+    return false;
+  }
```

Comment:
Although I understood it as I read the design doc, it would be helpful for future reference if we could note that the condition must include primary key (Only EQ) or index key (Either EQ or IS NULL) as JavaDoc comments.

Reply:
Right, it would be helpful. Added in be90064. PTAL!